src/schema/schema_validators.py

```diff
@@ -57,11 +58,127 @@ def validate_application_header_before_entity_create(normalized_entity_type, req
+                         f" {new_data_dict['uuid']} for DOI since it contains unpublished Datasets.")
+
+"""
+Validate the DOI parameters are presented as a pair during creation or modification.
+Even if one is populated already, disallow setting the other, so the data is consciously synced.
+Verify the values are compatible with each other.
+"""
+def verify_DOI_pair(property_key, normalized_entity_type, request, existing_data_dict, new_data_dict):
+    # Disallow providing one DOI parameter but not the other
+    if ('doi_url' in new_data_dict and 'registered_doi' not in new_data_dict) or \
+       ('doi_url' not in new_data_dict and 'registered_doi' in new_data_dict):
+        raise ValueError(   f"The properties 'doi_url' and 'registered_doi' must both be set in the same operation.")
+    # Since both DOI parameters are present, make sure neither is the empty string
+    if new_data_dict['doi_url'] == '' or new_data_dict['registered_doi'] == '':
+        raise ValueError(   f"The properties 'doi_url' and 'registered_doi' cannot be empty, when specified.")
+    # Check if doi_url matches registered_doi with the expected prefix
+    if new_data_dict['doi_url'] != SchemaConstants.EXPECTED_DOI_PREFIX_WITH_SLASH + new_data_dict['registered_doi']:
+        raise ValueError(   f"The 'doi_url' property should match the 'registered_doi' property, after"
+                            f" the prefix {SchemaConstants.EXPECTED_DOI_PREFIX_WITH_SLASH}.")
+"""
+Validate every entity in a list is of entity_type accepted
+
+Parameters
+----------
+property_key : str
+    The target property key
+normalized_type : str
+    Submission
+request: Flask request object
+    The instance of Flask request passed in from application request
+existing_data_dict : dict
+    A dictionary that contains all existing entity properties
+new_data_dict : dict
+    The json data in request body, already after the regular validations
+"""
+def collection_entities_are_existing_datasets(property_key, normalized_entity_type, request, existing_data_dict, new_data_dict):
+    # `dataset_uuids` is required for creating a Collection
+    # Verify each UUID specified exists in the uuid-api, exists in Neo4j, and is for a Dataset before
+    # proceeding with creation of Collection.
+    bad_dataset_uuids = []
+    for dataset_uuid in new_data_dict['dataset_uuids']:
+        try:
+            ## The following code duplicates some functionality existing in app.py, in
+            ## query_target_entity(), which also deals with caching. In the future, the
+            ## validation logic shared by this file and app.py should become a utility
+            ## module, shared by validators as well as app.py.  But for now, the code
+            ## is repeated for the following.
+
+            # Get cached ids if exist otherwise retrieve from UUID-API. Expect an
+            # Exception to be raised if not found.
+            dataset_uuid_entity = schema_manager.get_hubmap_ids(id=dataset_uuid)
```

Comment:
When the uuid doesn't exist in UUID MySQL, there's no need to go further and look up in Neo4j.

Reply:
I've encountered (and caused!) cases where this occurs on Dev, though I know it shouldn't in Prod.
If double-checking is bad enough to remove a check, I believe the UUID check would go, because I want the Neo4j response to check the entity type.
But I think the extra check provides some sanity during development, and might catch a stray case on other tiers. Do you think it is costly enough or confusing enough it should really come out, @yuanzhou?

Reply:
Understood. I'm good with your implementation.